### ISSUE TITLE
Psalm turning a variable's type to mixed for no reason?

### ISSUE THREAD

**Issue by ShadauxCat** [vimeo/psalm#5067]
So I can't really explain what's happening here. The following breaks with an error about the type of $foo->bar->baz being mixed:

```php
<?php
/**
 * @param object{
 *    bar?: object{
 *      baz?: string
 *   }
 * }|null $foo
 * @return void
 */
function test($foo)
{
	if (isset($foo) && isset($foo->bar->baz) && strlen($foo->bar->baz) > 0)
	{
		echo "" . $foo->bar->baz;
	}
}
```

Seems to me the type here is clearly string, since the isset() eliminates undefined/null. Weirdly, if I change it to this:

```php
<?php
/**
 * @param object{
 *    bar?: object{
 *      baz?: string
 *   }
 * }|null $foo
 * @return void
 */
function test($foo)
{
	if (isset($foo) && isset($foo->bar->baz))
	{
		echo "" . $foo->bar->baz;
	}
}
```

Then the echo line doesn't flag an error, even though it DOES flag one in the first example. 

Likewise changing it to this:

```php
<?php
/**
 * @param object{
 *    bar?: object{
 *      baz?: string
 *   }
 * }|null $foo
 * @return void
 */
function test($foo)
{
        if (isset($foo) && isset($foo->bar->baz))
        {
                if(strlen($foo->bar->baz) > 0)
                {
                                echo "" . $foo->bar->baz;
                }
        }
}
```

...also removes the errors. Seems like psalm is choking on long sequences of &&-joined conditionals.

**Comment by ShadauxCat** [comment]
https://psalm.dev/r/0d1ca98421
